### PR TITLE
Variable font support

### PR DESCRIPTION
**Pull request by ramezgerges**
## Variable Font Support

Adds OpenType variable font APIs to SkiaSharp and HarfBuzzSharp, enabling runtime font variation axis control, typeface cloning with custom design positions, and full axis/instance querying.

Related skia PR: https://github.com/mono/skia/pull/185

### New Types

| Type | Description |
|------|-------------|
| `SKFourByteTag` | Four-byte tag struct wrapping Skia's `SkFourByteTag`. Parse, ToString, implicit uint conversion. |
| `SKFontVariationAxis` | Generated struct: `Tag`, `Min`, `Default`, `Max`, `IsHidden` |
| `SKFontVariationPositionCoordinate` | Generated struct: `Axis` (tag), `Value` (float) |
| `SKFontArguments` | ref struct: `VariationDesignPosition`, `CollectionIndex` — accepts spans, arrays, stackalloc |

### New `SKTypeface` APIs

| API | Description |
|-----|-------------|
| `VariationDesignParameters` | Property → `SKFontVariationAxis[]` — all axes the font supports |
| `VariationDesignParameterCount` | Property → `int` — axis count for span pre-allocation |
| `GetVariationDesignParameters(Span<>)` | Span overload — allocation-free |
| `VariationDesignPosition` | Property → `SKFontVariationPositionCoordinate[]` — current axis values |
| `VariationDesignPositionCount` | Property → `int` |
| `GetVariationDesignPosition(Span<>)` | Span overload — allocation-free |
| `Clone(ReadOnlySpan<SKFontVariationPositionCoordinate>)` | Clone typeface with new variation position |
| `Clone(SKFontArguments)` | Clone with full control (position, collection index) |

### New HarfBuzzSharp `Face` APIs

| API | Description |
|-----|-------------|
| `HasVariationData` | Whether font has `fvar` table |
| `VariationAxisCount` | Axis count property |
| `VariationAxisInfos` | Property → `OpenTypeVarAxisInfo[]` |
| `GetVariationAxisInfos(Span<>)` | Span overload |
| `TryFindVariationAxis(Tag, out info)` | Find axis by tag |
| `NamedInstanceCount` | Named instance count property |
| `GetNamedInstanceDesignCoordsCount(int)` | Coord count for a named instance |
| `GetNamedInstanceDesignCoords(int)` | Design coords array |
| `GetNamedInstanceDesignCoords(int, Span<float>)` | Span overload |
| `GetNamedInstanceSubfamilyNameId(int)` | Name table ID for subfamily |
| `GetNamedInstancePostScriptNameId(int)` | Name table ID for PostScript name |

### New HarfBuzzSharp `Font` APIs

| API | Description |
|-----|-------------|
| `SetVariations(ReadOnlySpan<Variation>)` | Set variation values by tag |
| `SetVariationCoordsDesign(ReadOnlySpan<float>)` | Set design-space coordinates |
| `SetVariationCoordsNormalized(ReadOnlySpan<int>)` | Set normalized coordinates |
| `VariationCoordsNormalized` | Property → `int[]` — current normalized coords |
| `GetVariationCoordsNormalized(Span<int>)` | Span overload |
| `SetVariationNamedInstance(int)` | Select a named instance |

### Usage — Array-based

```csharp
using var data = SKData.Create(File.OpenRead("Inter.ttf"));
using var typeface = SKTypeface.FromData(data);

// Query axes
foreach (var axis in typeface.VariationDesignParameters)
    Console.WriteLine($"{axis.Tag}: {axis.Min} .. {axis.Default} .. {axis.Max}");
// Output: wght: 100 .. 400 .. 900

// Clone with bold weight
var position = new[] {
    new SKFontVariationPositionCoordinate {
        Axis = SKFourByteTag.Parse("wght"),
        Value = 900
    }
};
using var bold = typeface.Clone(position);
```

### Usage — Allocation-free with Span

```csharp
// Pre-allocate based on count
Span<SKFontVariationAxis> axes = stackalloc SKFontVariationAxis[typeface.VariationDesignParameterCount];
typeface.GetVariationDesignParameters(axes);

// Clone with stackalloc
Span<SKFontVariationPositionCoordinate> pos = stackalloc SKFontVariationPositionCoordinate[] {
    new() { Axis = SKFourByteTag.Parse("wght"), Value = 900 }
};
using var bold = typeface.Clone(pos);
```

### Console output
```text
Axes for Inter:
  opsz: 14 .. 14 .. 32
  wght: 100 .. 400 .. 900

Tag round-trip: wght → 0x77676874 → wght
```

### Visual comparison

| Base (wght: 400) | Bold (wght: 900) |
|---|---|
| Base weight | Bold weight |

### C API changes (companion PR in mono/skia)

- `sk_fourbytetag_t` typedef matching `SkFourByteTag`
- `sk_fontarguments_variation_axis_t` and `sk_fontarguments_variation_position_coordinate_t` structs
- `sk_typeface_get_variation_design_parameters()`, `sk_typeface_get_variation_design_position()`, `sk_typeface_clone_with_arguments()`
- `DEF_MAP` for layout-compatible struct conversion, `static_assert` in `sk_structs.cpp`

### Tests

151 tests covering:
- Exact axis values (tag, min, default, max, isHidden) for Distortable.ttf
- `SKFourByteTag`: Parse, ToString, char constructor, implicit conversion, equality, edge cases (null, empty, short, long strings), known tag hex values, native round-trip
- Span vs property equivalence for all overloads
- Static font graceful handling (empty arrays, zero counts)
- Clone preserves exact design position through native round-trip
- Rendering tests: different weights produce different pixels, ink coverage varies, consistent clones match
- HarfBuzz: axis queries, named instances, normalized coordinates, variation setting
- Negative index validation on all instance methods

### Gallery sample

Variable Fonts sample using Inter — weight and optical size sliders with weight spectrum comparison strip.

### CODE REVIEW

**Comment by mattleibow** [comment]
Hey @ramezgerges, quick question: did you run `pwsh ./utils/generate.ps1` to get this diff or was it something else?

Since we never got a chance to update the harfbuzz API bindings to the newer HarfBuzz version, I have been just bumping the underlying harfbuzz DEPS version and not re-generating the API. The changes now are so big that the work is a bit harder to do. So many weird types and things. What I usually do is drop back to 2.8.2 and the run the generator.

```pwsh
# In externals/skia/DEPS, revert harfbuzz back to the version we last generated from:
# Change: 2b3631a866b3077d9d675caa4ec9010b342b5a7c
# To:     63e15eac4f443fa53565d1e4fb9611cdd7814f28

dotnet cake --target=git-sync-deps
pwsh ./utils/generate.ps1
```

This will show us exactly what needs updating in binding/libHarfBuzzSharp.json so we can make sure the API is usable and not just direct mapping without the human touch.

There is a goal at some point to actually expose more, but never got to it. The version of harfbuzz being built is much later (still a bit old now) but not actually 2.8.2 - I think 8.x at least. There is a dream to one day bump harfbuzz and regen the full API and then be amazed.

**Comment by ramezgerges** [comment]
@mattleibow Yes, I ran the script for harfbuzz and took only the relevant parts. The PR does contain changes to the json files.

**Comment by mattleibow** [comment]
Uploading comparison screenshots for PR description.

| Base (wght: 400) | Bold (wght: 900) |
|---|---|

File: binding/HarfBuzzSharp/Font.cs [hunk 1]
```diff
@@ -291,6 +291,45 @@ public bool TryGetGlyphFromString (string s, out uint glyph)
...
+
+		public int[] GetVarCoordsNormalized ()
+		{
+			uint length;
+			var ptr = HarfBuzzApi.hb_font_get_var_coords_normalized (Handle, &length);
+			if (length == 0 || ptr == null)
+				return Array.Empty<int> ();
+
+			var coords = new int[length];
+			for (uint i = 0; i < length; i++)
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Font.cs]: Already addressed — the current code casts `length` to `int` for array allocation and the copy loop works correctly. The Span overload uses `Math.Min((int)length, coords.Length)` for safe bounds.

File: utils/SkiaSharpGenerator/BaseTool.cs [hunk 2]
```diff
@@ -50,7 +50,44 @@ protected void ParseSkiaHeaders()
...
+			foreach (var cp in clangSearchPaths)
+			{
+				if (Directory.Exists(cp))
+				{
+					Log?.LogVerbose($"Found Clang include folder: {cp}");
+					options.SystemIncludeFolders.Add(cp);
+					break;
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply utils/SkiaSharpGenerator/BaseTool.cs]: Pre-existing code from #3714 — not part of the variable font changes.

File: binding/HarfBuzzSharp/Face.cs [hunk 3]
```diff
@@ -85,6 +85,56 @@ public Blob ReferenceTable (Tag table) =>
...
+		{
+			var count = HarfBuzzApi.hb_ot_var_get_axis_count (Handle);
+			if (count == 0)
+				return Array.Empty<OpenTypeVarAxisInfo> ();
+
+			var axes = new OpenTypeVarAxisInfo[count];
+			fixed (OpenTypeVarAxisInfo* ptr = axes) {
+				HarfBuzzApi.hb_ot_var_get_axis_infos (Handle, 0, &count, ptr);
+			}
+			return axes;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Face.cs]: The array is pre-sized using `hb_ot_var_get_axis_count()` which returns the exact axis count. The `[inout] count` param to `hb_ot_var_get_axis_infos` will write at most that many entries, so the array is always fully populated. No trailing defaults possible.

File: binding/HarfBuzzSharp/Face.cs [hunk 4]
```diff
@@ -85,6 +85,56 @@ public Blob ReferenceTable (Tag table) =>
...
+			HarfBuzzApi.hb_ot_var_named_instance_get_postscript_name_id (Handle, (uint)instanceIndex);
+
+		public float[] GetNamedInstanceDesignCoords (int instanceIndex)
+		{
+			uint coordsLength = 0;
+			HarfBuzzApi.hb_ot_var_named_instance_get_design_coords (Handle, (uint)instanceIndex, &coordsLength, null);
+			if (coordsLength == 0)
+				return Array.Empty<float> ();
+
+			var coords = new float[coordsLength];
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Face.cs]: Already addressed — `GetNamedInstanceDesignCoords` now uses the return value from `hb_ot_var_named_instance_get_design_coords` (which is the total count per [HarfBuzz docs](https://harfbuzz.github.io/harfbuzz-hb-ot-var.html#hb-ot-var-named-instance-get-design-coords)) and casts to `int` for allocation.

File: binding/HarfBuzzSharp/Face.cs [hunk 5]
```diff
@@ -85,6 +85,56 @@ public Blob ReferenceTable (Tag table) =>
...
+		public OpenTypeNameId GetNamedInstanceSubfamilyNameId (int instanceIndex) =>
+			HarfBuzzApi.hb_ot_var_named_instance_get_subfamily_name_id (Handle, (uint)instanceIndex);
+
+		public OpenTypeNameId GetNamedInstancePostScriptNameId (int instanceIndex) =>
+			HarfBuzzApi.hb_ot_var_named_instance_get_postscript_name_id (Handle, (uint)instanceIndex);
+
+		public float[] GetNamedInstanceDesignCoords (int instanceIndex)
+		{
+			uint coordsLength = 0;
+			HarfBuzzApi.hb_ot_var_named_instance_get_design_coords (Handle, (uint)instanceIndex, &coordsLength, null);
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Face.cs]: Already addressed — all instance-index methods validate `instanceIndex < 0` and throw `ArgumentOutOfRangeException`. Tests cover all 5 methods.

File: binding/HarfBuzzSharp/Font.cs [hunk 6]
```diff
@@ -291,6 +291,45 @@ public bool TryGetGlyphFromString (string s, out uint glyph)
...
+				coords[i] = ptr[i];
+			return coords;
+		}
+
+		public void SetVarNamedInstance (int instanceIndex) =>
+			HarfBuzzApi.hb_font_set_var_named_instance (Handle, (uint)instanceIndex);
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Font.cs]: Already addressed — `SetVariationNamedInstance` (renamed from `SetVarNamedInstance`) checks `instanceIndex < 0`. Test `NegativeInstanceIndexThrowsForSetVariationNamedInstance` covers this.

File: tests/Tests/HarfBuzzSharp/HBFaceTest.cs [hunk 7]
```diff
@@ -1,11 +1,115 @@
...
 using Xunit;
 
 namespace HarfBuzzSharp.Tests
 {
 	public class HBFaceTest : HBTest
 	{
+		// Variable font helpers
+		private Face CreateVariableFace ()
+		{
+			var blob = Blob.FromFile (Path.Combine (PathToFonts, "Distortable.ttf"));
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/HarfBuzzSharp/HBFaceTest.cs]: Already addressed — `CreateVariableFace()` uses `using var blob`.

File: tests/Tests/HarfBuzzSharp/HBFontTest.cs [hunk 8]
```diff
@@ -1,12 +1,133 @@
...
 	public class HbFontTest : HBTest
 	{
+		// Variable font helpers
+		private (Face face, Font font) CreateVariableFontPair ()
+		{
+			var blob = Blob.FromFile (Path.Combine (PathToFonts, "Distortable.ttf"));
+			var face = new Face (blob, 0);
+			var font = new Font (face);
+			return (face, font);
+		}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/HarfBuzzSharp/HBFontTest.cs:18]: Already addressed — `CreateVariableFontPair()` uses `using var blob`.

File: binding/SkiaSharp/SKFontVariationAxis.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		/// Creates an OpenType tag from a four-character string (e.g. "wght", "wdth", "slnt").
+		/// </summary>
+		public static uint Make (string tag)
+		{
+			if (tag == null)
+				throw new ArgumentNullException (nameof (tag));
+			if (tag.Length != 4)
+				throw new ArgumentException ("Tag must be exactly 4 characters.", nameof (tag));
+
+			return ((uint)tag[0] << 24) | ((uint)tag[1] << 16) | ((uint)tag[2] << 8) | tag[3];
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/SkiaSharp/SKFontVariationAxis.cs]: Already addressed — `SKFontVariationTag` was replaced with `SKFourByteTag` struct, which uses `(byte)c1` casts in the char constructor (same pattern as `HarfBuzzSharp.Tag`). The old static class no longer exists.

File: samples/Gallery/Shared/Samples/VariableFontSample.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		for (int i = 0; i < weights.Length; i++)
+		{
+			var w = weights[i];
+			var cx = spacing * (i + 1);
+
+			var pos = new[]
+			{
+				new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("wght"), Value = w },
+				new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("opsz"), Value = opticalSize },
+			};
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply samples/Gallery/Shared/Samples/VariableFontSample.cs:112]: Intentionally kept simple — this is a gallery sample meant to clearly demonstrate the API, not production code. Caching would add complexity that obscures the font variation API usage.

File: binding/SkiaSharp/SKFourByteTag.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+	public SKFourByteTag (char c1, char c2, char c3, char c4)
+	{
+		value = (uint)(((byte)c1 << 24) | ((byte)c2 << 16) | ((byte)c3 << 8) | (byte)c4);
+	}
+
+	public static SKFourByteTag Parse (string tag)
+	{
+		if (string.IsNullOrEmpty (tag))
+			return new SKFourByteTag (0);
+
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/SkiaSharp/SKFourByteTag.cs]: Good catch — fixed. `Parse` now accepts `string?` to match the implementation that handles null.

File: binding/SkiaSharp/SKTypeface.cs [hunk 12]
```diff
@@ -362,6 +362,68 @@ public bool GetKerningPairAdjustments (ReadOnlySpan<ushort> glyphs, Span<int> ad
...
+			fixed (SKFontVariationPositionCoordinate* ptr = coordinates) {
+				return SkiaApi.sk_typeface_get_variation_design_position (Handle, ptr, coordinates.Length);
+			}
+		}
+
+		public SKTypeface Clone (ReadOnlySpan<SKFontVariationPositionCoordinate> position) =>
+			Clone (position, 0);
+
+		public SKTypeface Clone (ReadOnlySpan<SKFontVariationPositionCoordinate> position, int collectionIndex)
+		{
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/SkiaSharp/SKTypeface.cs:396]: The `Clone(span, int collectionIndex)` overload has been removed. `collectionIndex` is now set via `SKFontArguments.CollectionIndex` on the `Clone(SKFontArguments)` overload.

File: tests/Tests/SkiaSharp/SKVariableFontRenderingTest.cs [hunk 13]
```diff
@@ -0,0 +1,272 @@
+using System;
+using System.IO;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKVariableFontRenderingTest.cs]: Fixed — removed `using System.Linq`.

File: tests/Tests/SkiaSharp/SKVariableFontRenderingTest.cs [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+		};
+		using var cloned = staticTypeface.Clone (position);
+
+		if (cloned != null)
+		{
+			var clonedPixels = RenderTextToBitmapBytes (cloned, text, 32);
+			// For a static font, the output should be the same regardless of variation
+			Assert.False (PixelsDiffer (pixels, clonedPixels),
+				"Static font rendering should not change with variation parameters.");
+		}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKVariableFontRenderingTest.cs]: Good point — fixed. `Clone` on a static font returns a valid typeface (Skia handles this gracefully), so the test now asserts `Assert.NotNull(cloned)` and always runs the rendering comparison.

File: tests/Tests/HarfBuzzSharp/HBFontTest.cs [hunk 15]
```diff
@@ -1,12 +1,224 @@
...
+
+		[SkippableFact]
+		public void CanSetVarNamedInstance ()
+		{
+			var (face, font) = CreateVariableFontPair ();
+			using (face)
+			using (font) {
+				var count = face.NamedInstanceCount;
+				if (count == 0)
+					return;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/HarfBuzzSharp/HBFontTest.cs]: Already addressed — test now asserts `count > 0` (Distortable.ttf has 3 named instances).

File: tests/Tests/HarfBuzzSharp/HBFontTest.cs [hunk 16]
```diff
@@ -1,12 +1,224 @@
...
+		}
+
+		[SkippableFact]
+		public void CanSetMultipleVariationsSimultaneously ()
+		{
+			var (face, font) = CreateVariableFontPair ();
+			using (face)
+			using (font) {
+				var axes = face.VariationAxisInfos;
+				if (axes.Length < 1) return;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/HarfBuzzSharp/HBFontTest.cs:45]: Already addressed — test uses `Assert.NotEmpty(axes)`.

File: samples/Gallery/Shared/Samples/VariableFontSample.cs [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+		canvas.Clear(SKColors.White);
+
+		if (baseTypeface == null)
+			return;
+
+		var position = new[]
+		{
+			new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("wght"), Value = weight },
+			new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("opsz"), Value = opticalSize },
+		};
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply samples/Gallery/Shared/Samples/VariableFontSample.cs:62]: Intentionally kept simple for sample clarity — the `SKFourByteTag.Parse` calls and array allocations clearly show the API usage pattern. A production app would cache these.

File: samples/Gallery/Shared/Samples/VariableFontSample.cs [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+		DrawWeightSpectrum(canvas, width, height, typeface);
+	}
+
+	private void DrawWeightSpectrum(SKCanvas canvas, int width, int height, SKTypeface currentTypeface)
+	{
+		var spectrumY = height * 0.65f;
+		var spectrumText = "Aa";
+		var spectrumSize = Math.Max(20, textSize * 0.6f);
+		float[] weights = { 100, 200, 300, 400, 500, 600, 700, 800, 900 };
+		var spacing = width / (float)(weights.Length + 1);
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply samples/Gallery/Shared/Samples/VariableFontSample.cs]: Fixed — `weights` is now `private static readonly float[] SpectrumWeights`.

File: binding/SkiaSharp/SKTypeface.cs [hunk 19]
```diff
@@ -362,6 +362,72 @@ public bool GetKerningPairAdjustments (ReadOnlySpan<ushort> glyphs, Span<int> ad
...
+				return SkiaApi.sk_typeface_get_variation_design_position (Handle, ptr, coordinates.Length);
+			}
+		}
+
+		public SKTypeface Clone (ReadOnlySpan<SKFontVariationPositionCoordinate> position)
+		{
+			fixed (SKFontVariationPositionCoordinate* ptr = position) {
+				return GetObject (SkiaApi.sk_typeface_clone_with_arguments (Handle, ptr, position.Length, 0));
+			}
+		}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/SkiaSharp/SKTypeface.cs:400]: Fixed — PR description updated. The API is now `Clone(ReadOnlySpan<...>)` for the simple case and `Clone(SKFontArguments)` for full control (position + collectionIndex). The old `Clone(span, int)` overload was removed.

File: samples/Gallery/Shared/Samples/VariableFontSample.cs [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+			var pos = new[]
+			{
+				new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("wght"), Value = w },
+				new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("opsz"), Value = opticalSize },
+			};
+
+			using var tf = baseTypeface!.Clone(pos);
+			if (tf == null)
+				continue;
+
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply samples/Gallery/Shared/Samples/VariableFontSample.cs:117]: Kept as-is for sample clarity — the per-frame cloning clearly demonstrates the Clone API. A production app would cache typefaces keyed by weight/opsz.

File: binding/HarfBuzzSharp/Face.cs [hunk 21]
```diff
@@ -85,6 +85,100 @@ public Blob ReferenceTable (Tag table) =>
...
+
+		public float[] GetNamedInstanceDesignCoords (int instanceIndex)
+		{
+			if (instanceIndex < 0)
+				throw new ArgumentOutOfRangeException (nameof (instanceIndex));
+
+			// Return value is the total number of design coordinates
+			var totalCoords = (int)HarfBuzzApi.hb_ot_var_named_instance_get_design_coords (Handle, (uint)instanceIndex, null, null);
+			if (totalCoords == 0)
+				return Array.Empty<float> ();
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Face.cs:161]: The `coords_length` parameter is annotated `[inout][optional]` in the HarfBuzz docs — null is explicitly safe. The return value gives us the count regardless. This matches how the HarfBuzz docs describe the function: "Return value: the number of variation axes in the face". Our tests (`NamedInstanceDesignCoordsCountMatchesAxisCount`, `NamedInstanceDesignCoordsArrayLengthMatchesCount`) verify the returned count is correct.

File: tests/Tests/SkiaSharp/SKTypefaceTest.cs [hunk 22]
```diff
@@ -562,5 +562,464 @@ static IntPtr DoWork()
...
+			Assert.NotNull (typeface);
+
+			var position = new SKFontVariationPositionCoordinate[] {
+				new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse ("wght"), Value = 400 }
+			};
+
+			// Static fonts should handle this gracefully
+			using var cloned = typeface.Clone (position);
+			// Result may or may not be null depending on implementation
+		}
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply tests/Tests/SkiaSharp/SKTypefaceTest.cs:773]: Fixed — the test now asserts `Assert.NotNull(cloned)` and always runs the rendering comparison. Clone on a static font returns a valid typeface (Skia handles this gracefully), and the rendering should be identical regardless of variation parameters.

File: samples/Gallery/Shared/Samples/VariableFontSample.cs [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+		var position = new[]
+		{
+			new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("wght"), Value = weight },
+			new SKFontVariationPositionCoordinate { Axis = SKFourByteTag.Parse("opsz"), Value = opticalSize },
+		};
+
+		using var typeface = baseTypeface.Clone(position);
+		if (typeface == null)
+			return;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply samples/Gallery/Shared/Samples/VariableFontSample.cs:66]: Kept simple for sample clarity — the purpose is demonstrating the API, not optimizing rendering. The `weights` array was made `static readonly` (easy win), but caching tag parsing and coord buffers would obscure the API patterns we're trying to showcase. A production app would cache these.

File: binding/HarfBuzzSharp/Face.cs [hunk 24]
```diff
@@ -85,6 +85,100 @@ public Blob ReferenceTable (Tag table) =>
...
+			return HarfBuzzApi.hb_ot_var_named_instance_get_postscript_name_id (Handle, (uint)instanceIndex);
+		}
+
+		public int GetNamedInstanceDesignCoordsCount (int instanceIndex)
+		{
+			if (instanceIndex < 0)
+				throw new ArgumentOutOfRangeException (nameof (instanceIndex));
+
+			// Return value is the total number of design coordinates
+			return (int)HarfBuzzApi.hb_ot_var_named_instance_get_design_coords (Handle, (uint)instanceIndex, null, null);
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply binding/HarfBuzzSharp/Face.cs:150]: Per HarfBuzz docs, `coords_length` is `(inout) (optional)` — null is safe. We use the **return value** ("the number of variation axes in the face") as the count, not the `coords_length` out-param. This was actually a bug we caught and fixed during review — the original code incorrectly used the `[inout]` param (which stays 0 when input is 0) instead of the return value.
